monitoring/grafana/APIServerPriorityFairness.json

```diff
@@ -0,0 +1,845 @@
+          "interval": "",
+          "intervalFactor": 1,
+          "legendFormat": "{{priority_level}}",
+          "refId": "A"
+        }
+      ],
+      "title": "Shared concurrency limit by priority group",
+      "transparent": true,
+      "type": "stat"
+    },
+    {
+      "datasource": {
+        "default": true,
+        "type": "prometheus",
+        "uid": "prometheus-test"
+      },
+      "description": "",
+      "fieldConfig": {
+        "defaults": {
+          "color": {
+            "mode": "palette-classic"
+          },
+          "custom": {
+            "axisBorderShow": false,
+            "axisCenteredZero": false,
+            "axisColorMode": "text",
+            "axisGridShow": false,
+            "axisLabel": "",
+            "axisPlacement": "auto",
+            "barAlignment": 0,
+            "barWidthFactor": 0.6,
+            "drawStyle": "bars",
+            "fillOpacity": 25,
+            "gradientMode": "none",
+            "hideFrom": {
+              "legend": false,
+              "tooltip": false,
+              "viz": false
+            },
+            "insertNulls": false,
+            "lineInterpolation": "smooth",
+            "lineWidth": 3,
+            "pointSize": 3,
+            "scaleDistribution": {
+              "type": "linear"
+            },
+            "showPoints": "auto",
+            "spanNulls": false,
+            "stacking": {
+              "group": "A",
+              "mode": "none"
+            },
+            "thresholdsStyle": {
+              "mode": "off"
+            }
+          },
+          "decimals": 1,
+          "mappings": [],
+          "thresholds": {
+            "mode": "absolute",
+            "steps": [
+              {
+                "color": "green",
+                "value": null
+              },
+              {
+                "color": "red",
+                "value": 80
+              }
+            ]
+          },
+          "unit": "percent"
+        },
+        "overrides": []
+      },
+      "gridPos": {
+        "h": 11,
+        "w": 24,
+        "x": 0,
+        "y": 16
+      },
+      "id": 47,
+      "maxDataPoints": 500,
+      "options": {
+        "legend": {
+          "calcs": [
+            "lastNotNull",
+            "max"
+          ],
+          "displayMode": "table",
+          "placement": "right",
+          "showLegend": true
+        },
+        "tooltip": {
+          "mode": "multi",
+          "sort": "desc"
+        }
+      },
+      "targets": [
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "prometheus-test"
+          },
+          "editorMode": "code",
+          "exemplar": false,
+          "expr": "100 * max by (priority_level)(apiserver_flowcontrol_current_executing_requests{job=\"$APIService\", namespace=\"$SpaceNS\"}) / max by (priority_level) (apiserver_flowcontrol_request_concurrency_limit{job=\"$APIService\", namespace=\"$SpaceNS\"}) ",
```

Comment:
This one has an even worse subsampling problem. Its description in https://kubernetes.io/docs/reference/instrumentation/metrics/ is as follows.

> Number of requests in initial (for a WATCH) or any (for a non-WATCH) execution stage in the API Priority and Fairness subsystem.

So it is a gauge that just reports the current number, not a max or average or anything over a recent period of time.

Remember also that some priority levels have a concurrency limit of zero, which makes for a poor denominator in a division operation.

Again, there are metrics that take other approaches. One example is the `apiserver_flowcontrol_priority_level_seat_utilization` histogram. Its description is being updated in https://github.com/kubernetes/kubernetes/pull/126642 , which extends the description to say the following.

> Observations, at the end of every nanosecond, of utilization of seats for any stage of execution (but only initial stage for WATCHes); denominator is current seat limit or max(1, round(server seat limit/10))

However, that only covers requests that are APF has released for execution, not the ones being held in queues.

The `apiserver_flowcontrol_priority_level_request_utilization` histogram covers both, but in terms of requests rather than seats. https://github.com/kubernetes/kubernetes/pull/126642 is also updating its description, to the following.

> Observations, at the end of every nanosecond, of number of requests waiting or in any stage of execution (but only initial stage for WATCHes); denominator for waiting is max(1, QueueLengthLimit) X max(1, DesiredNumQueues), for executing is current seat limit if that is not zero otherwise max(1, round(server seat limit/10))

Another gauge, `apiserver_flowcontrol_demand_seats_high_watermark`, takes both waiting and executing requests into account, albeit with the inherent sampling problem. Its description is as follows.

> High watermark, over last adjustment period, of demand_seats.

You can see "demand" defined at https://github.com/kubernetes/enhancements/blob/e55a133b401e66bf18363b4de14f30578819eecf/keps/sig-api-machinery/1040-priority-and-fairness/README.md?plain=1#L871-L874 .

There are also `_average`, `_stdev`, and `_smoothed` variants of that gauge.  There is also a `apiserver_flowcontrol_demand_seats` histogram with `_buckets`, `_count`, and `_sum`; this is another histogram of values observed every nanosecond and thus does not suffer the smapling problem.